File: Doubley_Linked_list.cpp
#include<bits/stdc++.h>
#define        ll     long long int
using namespace std;

class node
{
public:
    int data;
    node *next;
    node *prev;
};

class linked_list
{
private:
    node *head,*tail;
public:
    int siz=0;
    linked_list()          //default constructor
    {
        head = NULL;
        tail = NULL;
    }

    void append(int n)
    {
        node *tmp = new node;
        tmp->data = n;
        tmp->next = NULL;
        tmp->prev = NULL;

        if(head == NULL)
        {
            head = tmp;
            tail = tmp;
        }
        else
        {
            tmp->prev=tail;
            tail->next = tmp;
            tail = tail->next;
        }
        siz++;
    }

    void prepend(int n)
    {
        node *tmp = new node;
        tmp->data = n;
        tmp->next = NULL;
        tmp->prev=NULL;

        if(head == NULL)
        {
            head = tmp;
            tail = tmp;
        }
        else
        {
            head->prev=tmp;
            tmp->next = head;
            head=tmp;
        }
        siz++;
    }


    void insert_at_nth(int n,int pos)
    {
        node *tmp = new node;
        tmp->data = n;
        tmp->next = NULL;
        tmp->prev=NULL;

        if(head == NULL)
        {
            head = tmp;
            tail = tmp;
            siz++;
        }
        else if(pos>siz)
        {
            append(n);
        }
        else
        {
            if(pos==1)
            {
                head->prev=tmp;
                tmp->next=head;
                head=tmp;
                siz++;
                return;
            }
            int x=1;
            node *v=head;
            while(v->next!=NULL)
            {
                if(x==pos-1)
                {
                    tmp->next=v->next;
                    tmp->prev=v->next->prev;
                    v->next=tmp;
                    tmp->next->prev=tmp;
                }
                x++;
                v=v->next;
            }
            siz++;
        }
    }


    void deleteNode(int n)
    {
        if(head == NULL)
        {
            cout<<"List is empty"<<endl;
            return;
        }
        if(head->data==n)   //if n is first node
        {
            head=head->next;
            head->prev=NULL;
            siz--;
            return;
        }
        node *ptr=head;
        while(ptr->next!=NULL)
        {
            if(ptr->next->data==n)
            {
                ptr->next=ptr->next->next;
                ptr->next->prev=ptr;
                if(ptr->next==NULL)
                {
                    tail=ptr;
                }
                siz--;
                return;
            }
            ptr=ptr->next;
        }
    }


    void printList()
    {
        node *tmp=new node;
        tmp=head;
        while(tmp->next!=NULL)
        {
            cout<<tmp->data<<" ";
            tmp=tmp->next;
        }
        cout<<tmp->data<<endl;
    }

    void printListReverse()
    {
        node *tmp=new node;
        tmp=tail;
        while(tmp->prev!=NULL)
        {
            cout<<tmp->data<<" ";
            tmp=tmp->prev;
        }
        cout<<tmp->data<<endl;
    }
};

int main()
{
    linked_list a;
    a.append(1);
    a.append(2);
    a.append(3);
    a.append(4);
    a.append(5);
    a.prepend(10);
    a.prepend(77);
    a.printList();
    a.insert_at_nth(111,8);
    a.printList();
    cout<<a.siz<<endl;
    a.deleteNode(77);
    a.printList();
    cout<<a.siz<<endl;
    a.deleteNode(5);
    a.printList();
    cout<<a.siz<<endl;
    a.printListReverse();
    return 0;
}

File: Doubley_Linked_list.py
class Node:
    def __init__(self,value):  #passing the value in Node
        self.next=None
        self.prev = None
        self.val=value


class doubleLinkedList:
    def __init__(self):
        self.head=None
        self.tail=None
        self.size=0

    def add_inThe_back(self,val):
        node=Node(val)           #creating a new node with value val
        if self.tail is None:    #checking the list is empty or not
            self.head=node
            self.tail=node
            self.size+=1
        else:                      #if the node is not empty assingning the value at the back of the list
            self.tail.next=node
            node.prev=self.tail
            self.tail=node
            self.size+=1

    def add_inThe_front(self,val):
        node=Node(val)
        if self.tail is None:
            self.head=node
            self.tail=node
            self.size+=1
        else:
            node.next=self.head    #if the node is not empty assingning the value at the front of the list
            self.head.prev=node
            self.head=node
            self.size+=1

    def add_at_nth(self,val,pos):
        node=Node(val)
        if self.tail is None:      #if the list is empty
            self.head=node
            self.tail=node
            self.size+=1
        else:
            if pos==1:              #if given position is 1
                node.next=self.head
                self.head.prev=node
                self.head=node
                self.size+=1
            elif pos>self.size:    #if the given position is the last position
                self.add_inThe_back(val)
            else:                 #if the position is not 1
                x=1
                node1=self.head
                while node1 is not None:
                    if x==pos-1:
                        node.next=node1.next
                        node.prev=node1.next.prev
                        node1.next=node
                        node.next.prev=node
                    x+=1
                    node1=node1.next




    def __remove_node(self,node):
        if node.prev is None:
            self.head=node.next
        else:
            node.prev.next = node.next
        if node.next is None:
            self.tail=node.prev
        else:
            node.next.prev = node.prev
        self.size-=1

    def remove_first(self):
        if self.head is not None:
            self.__remove_node(self.head)

    def remove_last(self):
        if self.tail is not None:
            self.__remove_node(self.tail)


    def remove(self,value):
        node=self.head
        while node is not None:
            if node.val==value:
                self.__remove_node(node)
            node=node.next


    def __str__(self):
        vals=[]
        node=self.head
        while node is not None:
            vals.append(node.val)
            node=node.next
        return f"[{', '.join(str(val) for val in vals)}]"
    
    def back(self):
        return self.tail.val

    def front(self):
        return self.head.val
    
    

my_list=doubleLinkedList()
my_list.add_inThe_back(3)
my_list.add_inThe_back(8)
my_list.add_inThe_back(6)
my_list.add_inThe_front(5)
my_list.add_inThe_front(1)
my_list.remove_last()
print(my_list)
my_list.add_at_nth(11,6)
print(my_list)
